enzy_htp/preparation/constrained_docking.py

```diff
@@ -0,0 +1,564 @@
+"""Drive code for the constrained docking protocol in enzy_htp. This implementation makes use of RosettaLigand and the enzdes constraints
+to dock structures with constrained geometries. Users are meant to call the `constrained_docking()` method and none of the other functions
+found in this sub-module. Consult `constrained_docking()` for more information on the protocol and its settings. This function is used in 
+both enzy_htp/preparation/reactive_docking.py and enzy_htp/preparation/place_ligand.py with differing settings.
+
+Author: Chris Jurich <chris.jurich@vanderbilt.edu>
+Date: 2023-09-20
+"""
+import os
+from pathlib import Path
+from typing import List, Tuple
+from copy import deepcopy
+
+import pandas as pd
+import numpy as np
+
+from enzy_htp.core import file_system as fs
+from enzy_htp.core import _LOGGER
+from enzy_htp._interface.rosetta_interface import RosettaCst
+from enzy_htp.structure import PDBParser, Structure
+
+from enzy_htp import interface, config
+
+
+def constrained_docking(structure: Structure,
+                        reactants: List[str],
+                        param_files: List[str],
+                        constraints: List[RosettaCst],
+                        work_dir: str = None,
+                        use_cache: bool = False,
+                        rng_seed: int = 1996,
+                        n_struct: int = 100,
+                        minimize: bool = True) -> pd.DataFrame:
```

Comment:
Now I get the idea of the complete thing is not finished rn so the IOs are not finalize.

Just make sure in the final version, functions the can be use not just for EnzyRCD (e.g.: this constrained docking) should take and output /structure objects or anything **insensitative to the change of engine**. (or in extreme cases that you think this function will never have any need to use another engine)

Reply:
And for parts that only EnzyRCD uses, you can have them stored in a submodule of `/preparation`

Reply:
yes I will make sure everything is in `Structure()`/`Ligand()` objects. For now I just wanted to get it working.

Reply:
Cool! However, the idea of `/structure` is to only have them store/provide_access_to the data. Here are some idea about where functions should go:
- Addressing a high-level scientific problem (like docking or protonation) -> "the 4 high-level modules"
- Operation of the structure that is going to be repeatedly used to address different scientific problems -> `structure.structure_operation`
- Operation of the structure that is only going to be used by EnzyRCD -> `preparation.reactive_docking.xxx`